caffe2/utils/threadpool/pthreadpool-cpp.cc

```diff
@@ -57,8 +72,22 @@ void PThreadPool::run(
 size_t getDefaultNumThreads();
 
 PThreadPool* pthreadpool() {
-  static std::unique_ptr<PThreadPool> threadpool =
-      std::make_unique<PThreadPool>(getDefaultNumThreads());
+  static auto threadpool =
+    std::make_unique<PThreadPool>(getDefaultNumThreads());
+#ifndef WIN32
+  static std::once_flag flag;
+  std::call_once(flag, []() {
+    pthread_atfork(nullptr, nullptr, child_atfork);
+  });
+#endif
+  auto true_bool = true;
+  if (leak_corrupted_threadpool.compare_exchange_strong(true_bool, false)) {
```

Comment:
@robieta is there an easy way to measure how much this check affects perf and how frequently this function is called?

I also wonder if we are likely to race here, and if global bool would be sufficient here and perhaps just annotated with `C10_UNLIKELY`

But all that can probably be done in followup PR

Reply:
@malfet, thanks for your guidance with this PR!

`std::atomic<bool>` is probably an overkill, so I can change it here if you'd like. 
I had used it because I don't understand the thread-model of PyTorch well, and since `num_threads` in `ParallelOpenMP.cpp` is atomic, I wasn't sure if there could be a race condition here as well. 
It'd be great if you could clarify why we use an atomic variable for `num_threads` in `ParallelOpenMP.cpp`, because I couldn't figure out what race-condition could arise there. I searched old issues & PRs, but couldn't find a reference. Thanks!

Reply:
At the very least, I think a weak CEX may be good enough. I bet @swolchok would know off the top of his head though.

Reply:
Hello @ezyang, do you think it's possible for weak CEX to cause an issue on NUMA nodes? Thanks!

Reply:
https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange says that when weak has to go in a loop and strong wouldn't, strong is better. (Weak is allowed to fail spuriously.)

Reply:
Hello @swolchok, do you think there could be a race-condition here if `leak_corrupted_threadpool` wasn't atomic, but simply a bool, and `C10_UNLIKELY` was used for it instead?

I used an atomic type for `leak_corrupted_threadpool` here, just to be on the safe side, because I don't understand the PyTorch thread-model well. Thank you!

Reply:
I think this is already racy as written with the CAS, let alone if you use the corrupted thread bool. Ordering of events:

Thread 1 enters the CAS
Thread 2 enters the CAS
Thread 1 sees true and writes false
Thread 2 sees false
thread 1 calls threadpool.release()
Thread 2 executes threadpool.get()

bam, unsynchronized access to threadpool where one thread writes and one thread reads. Offhand I think you need either a lock or atomics on the threadpool pointer itself.

I don't see any restrictions on which threads may call `pthreadpool()` and it's called via general-looking primitives like parallel_for in ParallelNative.h.

Reply:
I think this is already racy as written with the CAS, let alone if you use a non-atomic bool. Ordering of events:

Thread 1 enters the CAS
Thread 2 enters the CAS
Thread 1 sees true and writes false
Thread 2 sees false
thread 1 calls threadpool.release()
Thread 2 executes threadpool.get()

bam, unsynchronized access to threadpool where one thread writes and one thread reads. Offhand I think you need either a lock (which works via CAS internally anyway) or atomics on the threadpool pointer itself.

I don't see any restrictions on which threads may call pthreadpool() and it's called via general-looking primitives like parallel_for in ParallelNative.h.

Reply:
Thanks for the clarification, @swolchok!

Now that I think of it, this race-condition won't arise because `leak_corrupted_threadpool` would be set to true only after a fork, when the process would have one thread (after fork). All the threads would always see `leak_corrupted_threadpool` as false.

Reply:
> I also wonder if we are likely to race here, and if global bool would be sufficient here and perhaps just annotated with `C10_UNLIKELY`
But all that can probably be done in followup PR

@malfet, I think you were right. Opened #55341 to make `leak_corrupted_threadpool` a global bool.

Reply:
Hello again, @swolchok!

A digression from this PR - can you please give me an example of the following:
_OpenMP does not guarantee that a single per-process intra-op thread pool is going to be used in the application. On the contrary, two different application or inter-op threads may use different OpenMP thread pools for intra-op work._

It'd have been great if the [wiki](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/cpu_threading_torchscript_inference.rst) exemplified it.

Thank you!

Reply:
I don't know much about OpenMP, but I imagine that it has its own thread pools that are separate from the stuff getting touched in this PR.

Reply:
Thanks for your response, @swolchok! Yes, it's a digression from this PR.